### DISCUSSION TITLE
Implementation of PaperQA2 for Scientific Papers

### DISCUSSION THREAD

**Discussion by anand-kamble** [open-webui/open-webui#5896] [Ideas]
I would like to implement [PaperQA2](https://github.com/Future-House/paper-qa) in the Open WebUI. Since PaperQA2 offers better RAG performance for scientific papers, the need for processing scientific papers is increasing day by day. If implemented correctly it will be very beneficial for multiple educational institutions hosting their own servers for students and industries with confidential documents and want to host their own LLM chatbots. 

I have already started working on this and have created my own [fork](https://github.com/anand-kamble/open-webui) which I am testing and developing.

**Comment by justinh-rahb** [comment]
Why not just create a function for it? No fork needed.

**Reply by anand-kamble** [reply]
I have added a new PaperQA app in FastAPI apps and hence a new route `/paperqa/chat`, and then I am using it by calling the API from `chat.svelte` file, which handles the chat requests. I am trying to do most of the response streaming similar to what has been done with ollama, so I do not have to tamper with the middleware or any other part of the project.

@justinh-rahb Please let me know if there is a more efficient way of doing this since PaperQA uses LiteLLM and I think we can utilize that directly with the WebUI.

**Reply by palani-credain** [reply]
Anand-kamble, I have a fastapi app which I want to integrate into open web ui chat. should I add a new route for my app and add just modify chat.svelte? I would really appreciate if you can give me high level steps that I need to follow and if possible share your chat.svelte and any other file that needs to be modified.

**Reply by flefevre** [reply]
Paperqa2 utilizes litellm.
At first, an integration over a separated installation of Paperqa2 and an access over a dedicated fast api could be a minimum way to help many scientific. Which really mandatory for us 
If I do understand the proposition.
In this version open webui will act a frontend, a hub.
Advantage is to manage access authorization outside openwebui.

But we will lose the functionality of 'knowledbase' of open webui since documentation will be managed outside. A second solution which is really more complex is to rewrite a lot of openwebui where we could have multiple strategy/workflows for embedding (tika, docling, paperqa) and so externalise them for in a 'workspace layer', allowing manager of 'open webui' workspace' to setup the embedding, the vector database, the query system...of a workspace linked to a scientific thema.

I do not know how to do it.this was my 2 cents from France 7 am in the morning.

What is sure, having a more accurate system for scientific paper is really important for research and innovation.

'"We use the lmi package for our LLM interface, which in turn uses litellm to support many LLM providers. You can adjust this easily to use any model supported by litellm"
 https://github.com/Future-House/paper-qa#:~:text=We%20use%20the%20lmi%20package%20for%20our%20LLM%20interface%2C%20which%20in%20turn%20uses%20litellm%20to%20support%20many%20LLM%20providers.%20You%20can%20adjust%20this%20easily%20to%20use%20any%20model%20supported%20by%20litellm

**Comment by keyhanimo** [comment]
Would love to see this implemented!

**Comment by flefevre** [comment]
it will be a major update for the Scientific community to be able to use PaperQA2 directly from Open Webui.
it is cool that is already compatible with LiteLLM and Zotero.
Even if i know it is a bad idea to fork openwebui, I do not know how to make a link between open webui and paperQA2.
perhaps @tjbck could give some advices and methodology to do it?

**Reply by thiswillbeyourgithub** [reply]
Hi. I made a script to sync my zotero collections to open-webui. You might be interested in it.

Here's the link: [OpenWebUI Knowledge Zotero Sync
](https://github.com/thiswillbeyourgithub/openwebui-knowledge-zotero-sync)

**Comment by anand-kamble** [comment]
@tjbck & @justinh-rahb  I would like you to take a look at my PaperQA implementation. I have merged my branch with the latest Open WebUI commit, please have a look and let me know if what I have done is correct and what should I do to make it a PR.

Here is the link to the comparison: https://github.com/open-webui/open-webui/compare/main...anand-kamble:open-webui:main?expand=1

**Comment by flefevre** [comment]
Did you get any advice back? I do not understand how to implement a function that will covers the use case of integrating Paperqa2, is it possible to have more information?

How will you integrate the zotero feature? Will it be possible to extend the concept of knowledge base of open webui to link it to a given zotero library?
Should I create a separate ticket for this idea?

**Reply by anand-kamble** [reply]
Hello @flefevre ,
I haven't heard back from anyone, but I did implement the PaperQA2 [here](https://github.com/anand-kamble/open-webui/) in Open WebUI.
For Zotero, I haven't looked into it but that might be a good use case. 
And if you can create a ticket for this it will be amazing.

**Reply by flefevre** [reply]
I will try your work during holidays. Are you able to pull the last openwebui development?

Justin gives also the advice to make a function, why not going to this direction?

It will be definitely good to have an architecture discussion to choose the best solution

**Reply by anand-kamble** [reply]
I have synced the `main` branch of my repo with open-webui `main` branch. 
It would be great if you try it, and let me know about any improvements that can be made.

Also, since I am new to open-webui architecture I have made functions named `sendPromptPaperQA` similar to `sendPromptOllama` to handle PaperQA. Which get the job done, but I am hesitant to push those into a Pull Request.
  

Happy Holidays! 
> I will be working on this during the holidays too. Cheers!

**Reply by flefevre** [reply]
Just to be sure since my English is not so good. I am not part of open webui team. I have just fan of it and user in my scientists team.

**Reply by thiswillbeyourgithub** [reply]
Hi. I made a script to sync my zotero collections to open-webui. You might be interested in it.

Here's the link: [OpenWebUI Knowledge Zotero Sync
](https://github.com/thiswillbeyourgithub/openwebui-knowledge-zotero-sync)

**Comment by vishaldwdi** [comment]
@flefevre @anand-kamble @justinh-rahb 

what is the progress update? has PaperQA2 already been integrated into  https://github.com/open-webui/open-webui ??

if yes, then why README has no mention of it?
if not, then what's the plan ahead?

whether if there is any plan to integrate Zotero as well?

**Reply by anand-kamble** [reply]
Hello @vishaldwdi,

Over the past 2 months, I did not get enough time to make this into a complete PR. It's almost complete, I only have to work on citations getting displayed in the front-end correctly.  

About Zotero, I haven't planned that; neither I have worked with Zotero before, but if you can share any useful information or any packages that can implemented it will be great.

**Reply by flefevre** [reply]
For Zotero, I do not think you have to work in this direction at present time. I do not know the position of open webui team, they will potentially implement the MCP https://github.com/open-webui/open-webui/discussions/7363 . So you will have access to several MCP such has https://github.com/gyger/mcp-pyzotero or https://github.com/kaliaboi/mcp-zotero

I think the focus should be only on the core of paperqa but there is no precise guidelines on how they want to interface with paperqa

**Reply by vishaldwdi** [reply]
@anand-kamble well zotero can be dealt later once MCP is sorted, now that you already did good amount of work, it would be nice to push for its completion. i appreciate it.

**Reply by thiswillbeyourgithub** [reply]
Hi. I made a script to sync my zotero collections to open-webui. You might be interested in it.

Here's the link: [OpenWebUI Knowledge Zotero Sync
](https://github.com/thiswillbeyourgithub/openwebui-knowledge-zotero-sync)

**Comment by sir3mat** [comment]
Updates?

**Comment by connerlambden** [comment]
Since this thread has been looking for a scientific paper search integration, one option that works today via MCP is [BGPT](https://github.com/connerlambden/bgpt-mcp). It searches a database of papers built from raw experimental data extracted from full-text studies — so you get structured fields like methods, results, sample sizes, and statistical analyses rather than just abstracts.

You can connect it as an MCP server:
```
npx bgpt-mcp
```
or via remote endpoints (`https://bgpt.pro/mcp/sse`). 50 free searches, no API key needed.

It complements PaperQA2 nicely — PaperQA2 excels at RAG over your own paper collection, while BGPT gives broad search across its indexed corpus with structured experimental data already extracted. Could be useful as a discovery layer before feeding papers into PaperQA2 for deeper Q&A.
